src/main/java/org/opensearch/geospatial/index/mapper/xypoint/XYPointParser.java

```diff
@@ -71,101 +78,158 @@ public static XYPoint parseXYPoint(Object value, final boolean ignoreZValue) thr
-                    }
-                    if (Y_PARAMETER.equals(field)) {
-                        subParser.nextToken();
-                        switch (subParser.currentToken()) {
-                            case VALUE_NUMBER:
-                            case VALUE_STRING:
-                                try {
-                                    y = subParser.doubleValue(TRUE);
-                                } catch (NumberFormatException numberFormatException) {
-                                    throw new OpenSearchParseException("[y] must be valid double value", numberFormatException);
-                                }
-                                break;
-                            default:
-                                throw new OpenSearchParseException("[y] must be a number");
-                        }
-                    }
-                }
+        switch (parser.currentToken()) {
+            case START_OBJECT:
+                parseXYPointObject(parser, point, ignoreZValue);
+                break;
+            case START_ARRAY:
+                parseXYPointArray(parser, point, ignoreZValue);
+                break;
+            case VALUE_STRING:
+                String val = parser.text();
+                point.resetFromString(val, ignoreZValue);
+                break;
+            default:
+                throw new OpenSearchParseException("xy_point expected");
+        }
+        return point;
+    }
+
+    private static XYPoint parseXYPointObject(final XContentParser parser, final XYPoint point, final boolean ignoreZValue)
+        throws IOException {
+        try (XContentSubParser subParser = new XContentSubParser(parser)) {
+            if (subParser.nextToken() != XContentParser.Token.FIELD_NAME) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, subParser.currentToken());
             }
-            if (Double.isNaN(x)) {
-                throw new OpenSearchParseException("field [{}] missing", X_PARAMETER);
+
+            String field = subParser.currentName();
+            if (X_PARAMETER.equals(field) || Y_PARAMETER.equals(field)) {
+                parseGeoPointObjectBasicFields(subParser, point);
+            } else if (GEOJSON_TYPE.equals(field) || GEOJSON_COORDS.equals(field)) {
+                parseGeoJsonFields(subParser, point, ignoreZValue);
+            } else {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
             }
-            if (Double.isNaN(y)) {
-                throw new OpenSearchParseException("field [{}] missing", Y_PARAMETER);
+
+            if (subParser.nextToken() != XContentParser.Token.END_OBJECT) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
             }
-            return point.reset(x, y);
+
+            return point;
         }
+    }
+
+    private static XYPoint parseGeoPointObjectBasicFields(final XContentParser parser, final XYPoint point) throws IOException {
```

Comment:
why there is Geo in the name of the function?

Reply:
Good catch. Thanks.